common/cuda_hip/factorization/lu_kernels.cpp

```diff
@@ -113,14 +113,17 @@ __global__ __launch_bounds__(default_block_size) void factorize(
     // for each lower triangular entry: eliminate with corresponding row
     for (auto lower_nz = row_begin; lower_nz < row_diag; lower_nz++) {
         const auto dep = cols[lower_nz];
-        // we can load the value before synchronizing because the following
-        // updates only go past the diagonal of the dependency row, i.e. at
-        // least column dep + 1
-        const auto val = vals[lower_nz];
         const auto diag_idx = diag_idxs[dep];
         const auto dep_end = row_ptrs[dep + 1];
         scheduler.wait(dep);
+        // We need to load vals after synchronize.
+        // the next lower_nz might be modified if the dep row has the same col
+        // as next lower_nz's col.
+        const auto val = vals[lower_nz];
```

Comment:
Not sure I follow - each warp only modifies memory locations belonging to its row, so there are only data races between threads of the same warp. So is this reordering actually necessary?

Reply:
What you essentially want to do is add another warp sync at the end of each iteration of the outer loop, to prevent the modification of previous loops racing with ready from the following loop iterations? I would prefer having that sync happen explicitly at the end of the loop than hidden inside the scheduler wait function.

Reply:
Exactly. It is to ensure that the modification of the same warp can be seen from the others when getting it.
Yes, it is what I mean in the PR description. We only need warp sync.
I thought the wait(dep) should also imply that because it make the memory visible in block at least.
I just move that rather than introducing the warp sync